allennlp/modules/similarity_functions/multiheaded.py

```diff
@@ -0,0 +1,114 @@
+    """
+    def __init__(self,
+                 num_heads: int,
+                 tensor_1_dim: int,
+                 tensor_1_projected_dim: int = None,
+                 tensor_2_dim: int = None,
+                 tensor_2_projected_dim: int = None,
+                 internal_similarity: SimilarityFunction = DotProductSimilarity()) -> None:
+        super(MultiHeadedSimilarity, self).__init__()
+        self._num_heads = num_heads
+        self._internal_similarity = internal_similarity
+        tensor_1_projected_dim = tensor_1_projected_dim or tensor_1_dim
+        tensor_2_dim = tensor_2_dim or tensor_1_dim
+        tensor_2_projected_dim = tensor_2_projected_dim or tensor_2_dim
+        if tensor_1_projected_dim % num_heads != 0:
+            raise ConfigurationError("Projected dimension not divisible by number of heads: %d, %d"
+                                     % (tensor_1_projected_dim, num_heads))
+        if tensor_2_projected_dim % num_heads != 0:
+            raise ConfigurationError("Projected dimension not divisible by number of heads: %d, %d"
+                                     % (tensor_2_projected_dim, num_heads))
+        self._tensor_1_projection = Parameter(torch.Tensor(tensor_1_dim, tensor_1_projected_dim))
+        self._tensor_2_projection = Parameter(torch.Tensor(tensor_2_dim, tensor_2_projected_dim))
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        torch.nn.init.xavier_uniform(self._tensor_1_projection)
+        torch.nn.init.xavier_uniform(self._tensor_2_projection)
+
+    @overrides
+    def forward(self, tensor_1: torch.Tensor, tensor_2: torch.Tensor) -> torch.Tensor:
+        projected_tensor_1 = torch.matmul(tensor_1, self._tensor_1_projection)
+        projected_tensor_2 = torch.matmul(tensor_2, self._tensor_2_projection)
+
+        # split_tensor_{1,2} are both tuples of length num_heads, where each element in the tuple
+        # is a tensor of shape (..., projected_dim / num_heads).
+        split_size = projected_tensor_1.size(-1) // self._num_heads
+        split_tensor_1 = projected_tensor_1.split(split_size, dim=-1)
+        split_size = projected_tensor_2.size(-1) // self._num_heads
+        split_tensor_2 = projected_tensor_2.split(split_size, dim=-1)
+
+        # We then stack the tensors, inserting a new dimension in the _second-to-last_ position,
+        # giving us a tensor of shape (..., num_heads, projected_dim / num_heads).
+        stacked_tensor_1 = torch.stack(split_tensor_1, dim=-2)
+        stacked_tensor_2 = torch.stack(split_tensor_2, dim=-2)
+
+        # And then we pass this off to our internal similarity function.  Because the similarity
+        # functions don't care what dimension their input has, and only look at the last dimension,
+        # we don't need to do anything special here.  It will just compute similarity on the
+        # (..., num_heads, *projected_dim / num_heads*) dimension, returning a tensor of shape
```

Comment:
Are you just trying to say that the similarity is computed along the projection dimension for each head? This `*`notation is a little confusing.

Reply:
Done.